### PR TITLE
Fix lizard snouts disappearing when wearing masks

### PR DESCRIPTION
**Pull request by SlamBamActionman**
## About the PR

Bugfix, forgot to make nose toggling when wearing mask species-dependent in the nose PR #25557. 

## Why / Balance

fear
image

## Technical details

Adds a "HideNoseOnEquip" to HumanoidAppearanceComponent, which is set to true on humans and dwarves. Nose hiding logic now checks for it in ClothingSystem.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl: 
- fix: Snouts and noses no longer disappear with toggled masks.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
Isnt the issue that the nose layers arent being restored when the mask is removed or toggled? Regardless of whether or not lizard snouts should toggle, an unmasked lizard (like the one pictured in the PR) should not be hiding the snout, regardless of the species.

**Comment by Kukutis96513** [comment]
Screenshot_1234

it's absolutely horrifying, thanks for fixing it!

**Comment by SlamBamActionman** [comment]
I've done some further investigation into this and it appears the issue is specifically this line which checks for the mask component: 

```
if (TryComp(item, out MaskComponent? mask) && TryComp(item, out ClothingComponent? clothing))
{
    if (clothing.EquippedPrefix != mask.EquippedPrefix)
    {
        shouldLayerShow = false;
        break;
    }
}
else
{
    shouldLayerShow = false;
    break;
}
```

The reason I think this is because I can not replicate the bug with hidden hair, so it'd have to be the code which is mask-specific.

Steps to recreate the bug:
1. Play a lizard/human with nose marking.
2. Put on a gas mask and toggle it pulled down.
3. /aghost and leave the area; distance/time is unreliable, but I got it to trigger sometimes when heading far away and waiting for a minute or two.
4. Return to the character; nose should now be missing.

My guess would be that the bug triggers when the character is loaded in, which is also why it's more common to find other players missing their nose than your own. Since markings are shown by default it suggests the ToggleVisualLayer function is run upon loading. Could it be that the MaskComponent isn't loaded by the time that the function is run, thus failing the if statement?

**Comment by SlamBamActionman** [comment]
@EmoGarbage404 I've found the cause behind it, but I am uncertain of how to solve it.

> Tl;dr: Async (I think?) functions sometime cause the "should snout be hidden" check to be made _before_ the mask is determined to be pulled down, whenever a mob is loaded in clientside.

When a mob gets far enough away from the client player, it gets unloaded. For some reason, part of that unloading is to run `OnMaskToggle` and set the mask's `equippedPrefix` variable to `null`. This equivalent to the mask getting put in the pulled-up state. 

When the mob later gets loaded back in by the client, there is a check run for all clothes to restore `equippedPrefix`es to whatever they were before, via `OnHandleState`. This sets the mask's `equippedPrefix` back to `"toggled"`, which means the mask is set to the pulled-down state. All is good, the mask is shown in the correct state.

The issue comes from how `OnGotEquipped` is run for all equipped items upon the mob being loaded (which I assume is because the mob is re-instantiated and clothes applied on it). `OnGotEquipped` is what controls the snout layer visuals, but sometimes it's run *before* `OnHandleState`. This means that the mask's prefix is read as null (i.e. pulled-up) when determining if the snout should be hidden.

An option would be to run the visuals check in `OnHandleState` as well, though that's basically running the code twice. It might work, but might also be a bit smelly?

**Comment by EmoGarbage404** [comment]
i think running it after the state is handled is fine.

**Comment by SlamBamActionman** [comment]
I'm struggling with implementing the fix a bit; all other events have so far included a clean way to get the wearer's uid*. `OnHandleState` and `OnGetState` don't include it. You could probably store it in the `ClothingComponentState` like the `equippedPrefix` is, but then you'd have to put and track the wearer in a `ClothingComponent` variable and that two-way connection feels a bit icky.

I'll do it if it's not bad practice, but seeing as the On_State events are very baseline robust toolbox stuff I don't see how else to get the info.

*The reason the wearer's uid is necessary is to check all the wearer's equipment for snout hiding when an item is unequipped/toggle, like if they are wearing a welding mask and a gas mask at the same time and removes one of them.

**Comment by Tayrtahn** [comment]
I'm not certain, but SharedContainerSystem's TryGetContainingContainer might give you what you need. Then you can use the container's Owner to get the uid. (container ECS when?)

**Comment by SlamBamActionman** [comment]
> I'm not certain, but SharedContainerSystem's TryGetContainingContainer might give you what you need. Then you can use the container's Owner to get the uid. (container ECS when?)

This appeared to be the solution I was looking for! Thanks a ton! 

I can no longer reproduce the bug with this fix (not with human noses either), so hopefully this should have solved it.

File: Content.Shared/Clothing/EntitySystems/ClothingSystem.cs [hunk 1]
```diff
@@ -176,6 +174,15 @@ private void OnUnequipDoAfter(Entity<ClothingComponent> ent, ref ClothingUnequip
...
 
+    private void CheckEquipmentForLayerHide(string slot, EntityUid equipment, EntityUid equipee)
+    {
+        if ((new string[] { "head" }).Contains(slot) && _tagSystem.HasTag(equipment, HairTag))
+            ToggleVisualLayer(equipee, HumanoidVisualLayers.Hair, HairTag);
+        if ((new string[] { "mask", "head" }).Contains(slot) && _tagSystem.HasTag(equipment, NoseTag))
+            if (TryComp(equipee, out HumanoidAppearanceComponent? appearance) && appearance.HideNoseOnEquip)
+                ToggleVisualLayer(equipee, HumanoidVisualLayers.Snout, NoseTag);
+    }
+
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/EntitySystems/ClothingSystem.cs:185]: This isn't your fault but this code is extremely ass:
- it should be using the enums not strings.
- It shouldn't be re-allocating a string array every single time this is called.
- It shouldn't be hardcoded to specific layers and copy-pasted, it should just be generic between the 2 i.e. "HideNoseOnEquip" should be a bitmask of the slots.
- TagComponent should be re-used between them.
